sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp

```diff
@@ -512,4 +528,353 @@ cdef class EuclideanArgKmin{{name_suffix}}(ArgKmin{{name_suffix}}):
+            free(self.heaps_indices_chunks)
+
+        if self.heaps_r_distances_chunks is not NULL:
+            free(self.heaps_r_distances_chunks)
+
+#######################
+    @final
+    cdef void _parallel_on_X_parallel_init(
+        self,
+        BaseEngine engine,
+        ITYPE_t thread_num,
+    ) nogil:
+        BaseDistancesReduction{{name_suffix}}._parallel_on_X_parallel_init(self, engine, thread_num)
+        engine._parallel_on_X_parallel_init(thread_num)
+
+    cdef void _parallel_on_X_init_chunk(
+        self,
+        BaseEngine engine,
+        ITYPE_t thread_num,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+    ) nogil:
+        # As this strategy is embarrassingly parallel, we can set each
+        # thread's heaps pointer to the proper position on the main heaps.
+        self.heaps_r_distances_chunks[thread_num] = &self.argkmin_distances[X_start, 0]
+        self.heaps_indices_chunks[thread_num] = &self.argkmin_indices[X_start, 0]
+
+        engine._parallel_on_X_init_chunk(thread_num, X_start, X_end)
+
+    cdef void _parallel_on_X_pre_compute_and_reduce_distances_on_chunks(
+        self,
+        BaseEngine engine,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+        ITYPE_t Y_start,
+        ITYPE_t Y_end,
+        ITYPE_t thread_num,
+    ) nogil:
+        BaseDistancesReduction{{name_suffix}}._parallel_on_X_pre_compute_and_reduce_distances_on_chunks(
+            self, engine,
+            X_start, X_end,
+            Y_start, Y_end,
+            thread_num,
+        )
+        engine._parallel_on_X_pre_compute_and_reduce_distances_on_chunks(
+            X_start, X_end, Y_start, Y_end, thread_num,
+        )
+
+    cdef void _parallel_on_Y_pre_compute_and_reduce_distances_on_chunks(
+        self,
+        BaseEngine engine,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+        ITYPE_t Y_start,
+        ITYPE_t Y_end,
+        ITYPE_t thread_num,
+    ) nogil:
+        BaseDistancesReduction{{name_suffix}}._parallel_on_Y_pre_compute_and_reduce_distances_on_chunks(
+            self, engine,
+            X_start, X_end,
+            Y_start, Y_end,
+            thread_num,
+        )
+        engine._parallel_on_Y_pre_compute_and_reduce_distances_on_chunks(
+            X_start, X_end, Y_start, Y_end, thread_num
+        )
+    cdef void _compute_and_reduce_distances_on_chunks(
+        self,
+        BaseEngine engine,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+        ITYPE_t Y_start,
+        ITYPE_t Y_end,
+        ITYPE_t thread_num,
+    ) nogil:
+        cdef:
+            ITYPE_t i, j
+            DTYPE_t val
+            ITYPE_t n_X = X_end - X_start
+            ITYPE_t n_Y = Y_end - Y_start
+            DTYPE_t * heaps_r_distances = self.heaps_r_distances_chunks[thread_num]
+            ITYPE_t * heaps_indices = self.heaps_indices_chunks[thread_num]
+
+            DTYPE_t * dist_middle_terms = engine._compute_dist_middle_terms(
```

Comment:
This implementation uses `dist_middle_terms` and therefore is specific to the Euclidean case, but `ArgKmin` is not. For the generic case, you still need to call `self.datasets_pair.surrogate_dist()`.

One way I see is to migrate the logic of this function to `_compute_and_reduce_distances_on_chunks()` of BaseEngine and its subclasses. 

Also, I would avoid calling `_compute_pair_distance` for each `i, j` since this would probably lead to a performance decrease because of the overhead, as in https://github.com/scikit-learn/scikit-learn/pull/25170.

All in all, you'll end up with something like:

```python
# _argkmin.pyx.tp
 cdef void _compute_and_reduce_distances_on_chunks(
     ...
) nogil:
    cdef:
        DTYPE_t * heaps_r_distances = self.heaps_r_distances_chunks[thread_num]
        ITYPE_t * heaps_indices = self.heaps_indices_chunks[thread_num]
        
        engine._compute_and_reduce_distances_on_chunks(
            self.datasets_pair,  # for generic computation
            heaps_r_distances,
            heaps_indices,
            X_start,
            X_end,
            Y_start,
            Y_end,
            self.k,
        )
```

```python        
# _engines.pyx.tp
# BaseEngine
cdef void _compute_and_reduce_distances_on_chunks(
    DatasetsPair datasets_pair,
    DTYPE_t *heaps_r_distances,
    ITYPE_t *heaps_indices,
    ITYPE_t X_start,
    ITYPE_t X_end,
    ITYPE_t Y_start,
    ITYPE_t Y_end,
    ITYPE_t k,
) nogil:
    cdef:
        ITYPE_t i, j
        ITYPE_t n_X = X_end - X_start
        ITYPE_t n_Y = Y_end - Y_start
    for i in range(n_X):
        for j in range(n_Y):
            heap_push(
                values=heaps_r_distances + i * k,
                indices=heaps_indices + i * k,
                size=k,
                val=datasets_pair.surrogate_dist(X_start + i, Y_start + j),
                val_idx=j + Y_start,
            )
```

```python
# _engines.pyx.tp
# EuclideanEngine
cdef void _compute_and_reduce_distances_on_chunks(
    DatasetsPair datasets_pair,
    DTYPE_t *heaps_r_distances,
    ITYPE_t *heaps_indices,
    ITYPE_t X_start,
    ITYPE_t X_end,
    ITYPE_t Y_start,
    ITYPE_t Y_end,
    ITYPE_t k,
) nogil:
    cdef:
        ITYPE_t i, j
        ITYPE_t n_X = X_end - X_start
        ITYPE_t n_Y = Y_end - Y_start
        DTYPE_t val
        dist_middle_terms = self._compute_dist_middle_terms(
            X_start, X_end, Y_start, Y_end, thread_num
        )
    for i in range(n_X):
        for j in range(n_Y):
           
            val = max(0,
                self.X_norm_squared[X_start+i]
                + dist_middle_terms[i * n_Y + j]
                + self.Y_norm_squared[Y_start+j]
            )
            heap_push(
                values=heaps_r_distances + i * k,
                indices=heaps_indices + i * k,
                size=k,
                val=val,
                val_idx=j + Y_start,
            )
```

What do you think?

Reply:
I have implemented this suggestion for the most part, however I opted to use a function pointer for the surrogate distance (setting to `NULL` in the euclidean specialization) instead of passing `DatasetsPair` that way we neither have to specialize to 32/64 bit explicitly for the `EuclideanEngine` class and avoid further entangling `DatasetsPair` with the engines.